src/guest_config/src/guest_config_compiler_bin.rs

```diff
@@ -0,0 +1,366 @@
+                            format!("{}-{}", args.input_file, elapsed.as_secs(),)
+                    }
+                    Err(err) => panic!("Critical system error retrieving clock time - [{:?}]", err),
+                }
+            } else {
+                template_output_file_path = args.output_file.to_string();
+            }
+
+            match std::fs::write(&template_output_file_path, template_json_string) {
+                Ok(_) => Ok(()),
+                Err(err) => Err(GuestConfigCompilerError::IOError(
+                    template_output_file_path,
+                    err,
+                )),
+            }
+        }
+        Err(err) => Err(GuestConfigCompilerError::TemplateConfiguration(err)),
+    }
+}
+
+fn main() {
+    let mut arg_parser = build_arg_parser();
+
+    if let Err(err) = arg_parser.parse_from_cmdline() {
+        eprintln!(
+            "Arguments parsing error: {} \n\nFor more information try --help.",
+            err
+        );
+        process::exit(EXIT_CODE_ERROR);
+    }
+
+    if arg_parser.arguments().flag_present("help") {
+        println!(
+            "{} v{}\n",
+            GUESTCONFIG_COMPILER_FILE_NAME, GUESTCONFIG_COMPILER_VERSION
+        );
+        println!("{}", arg_parser.formatted_help());
+        return;
+    }
+
+    let args = get_argument_values(arg_parser.arguments()).unwrap_or_else(|err| {
+        eprintln!("{:?} \n\nFor more information try --help.", err);
+        process::exit(EXIT_CODE_ERROR);
+    });
+
+    if let Err(err) = compile(&args) {
+        eprintln!("guestconfig compilation error: {:?}", err);
+        process::exit(EXIT_CODE_ERROR);
+    }
+
+    println!("Template successfully compiled into: {}", args.output_file);
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{fs, io};
+
+    use cpuid::Cpuid;
+    use logger::error;
+    use tempfile::Builder;
+    use vm_guest_config::GuestConfigurationError;
+
+    use super::{
+        build_arg_parser, compile, get_argument_values, Arguments, GuestConfigCompilerError,
+        GUESTCONFIG_DEFAULT_OUTPUT_FILENAME,
+    };
+    use crate::GUESTCONFIG_COMPILER_FILE_NAME;
+
+    #[test]
+    fn test_errors() {
+        let path = "nonexistent";
+        assert_eq!(
+            format!(
+                "{:?}",
+                GuestConfigCompilerError::IOError(
+                    String::from(path),
+                    io::Error::from_raw_os_error(2)
+                )
+            ),
+            r#"IOError("nonexistent", Os { code: 2, kind: NotFound, message: "No such file or directory" })"#
+        );
+        assert_eq!(
+            format!(
+                "{}",
+                GuestConfigCompilerError::TemplateConfiguration(GuestConfigurationError::CpuId)
+            ),
+            format!(
+                "{}",
+                r#"Error processing virtualized guest's configuration. See error [Failed to configure CPU features via CPUID]"#
+            )
+        );
```

Comment:
This can be simplified.
```suggestion
        assert_eq!(
            GuestConfigCompilerError::TemplateConfiguration(GuestConfigurationError::CpuId).to_string(),
            r#"Error processing virtualized guest's configuration. See error [Failed to configure CPU features via CPUID]"#
        );
```